### PR TITLE
New feature 'sys'.

### PR DESCRIPTION
**Pull request by enjoyyourcopepods**
Defined the new method 'sys' in lib/sequenceserver.rb. Added "require
'tempfile'" to lib/sequenceserver.rb for 'sys'. Created the
CommandFailed class in lib/sequenceserver/exceptions.rb for 'sys'.
Changed lib/sequenceserver/database.rb to make use of 'sys'.

Signed-off-by: Lawrence John Maynard <enjoyyourcopepods@gmx.com>

### CODE REVIEW

File: lib/sequenceserver.rb [hunk 1]
```diff
@@ -57,6 +58,83 @@ def init(config = {})
...
 
+    # 'sys' executes a shell command.
+    #
+    # 'sys' can write the stdout and/or stderr from a shell command to files, or
+    #  return these values.
+    #
+    # 'sys' can get from a failed shell command stdout, stderr, and exit status.
+    #
+    # Supply 'sys' with the shell command and optionally:
+    # A directory to change to for the duration of the execution of the shell
```
Comment (yeban) [inline lib/sequenceserver.rb]: Give the name of the corresponding keys (`directory: A directory to change to ...`).

File: lib/sequenceserver.rb [hunk 2]
```diff
@@ -57,6 +58,83 @@ def init(config = {})
...
+    # A directory to change the PATH environment variable to for the duration of the
+    # execution of the shell command.
+    # A path to a file to write stdout to.
+    # A path to a file to write stderr to.
+    #
+    # Usage:
+    #
+    # stdout, stderr = sys(command, :directory => '/path/to/directory',
+    # :path => '/path/to/directory', :stdout_file => '/path/to/stdout_file',
+    # :stderr_file => '/path/to/stderr_file'
```
Comment (yeban) [inline lib/sequenceserver.rb]: Invoking this way doesn't return anything, right?

File: lib/sequenceserver.rb [hunk 3]
```diff
@@ -57,6 +58,83 @@ def init(config = {})
...
+        system("#{command} 1>#{temp_stdout_file.path} 2>#{temp_stderr_file.path}")
+      end
+
+      unless $CHILD_STATUS.success?
+        raise CommandFailed.new(temp_stdout_file.read, temp_stderr_file.read, $CHILD_STATUS.exitstatus)
+      end
+
+      # Write stdout and/or stderr to files, if paths for them were given.
+      # If a full path was given for an output file, write the contents of
+      # the temporary file to this path. If the path given does not
```
Comment (yeban) [inline lib/sequenceserver.rb]: We don't write the tempfiles to the given location, we move it (it's a feature, hence worth highlighting).

File: lib/sequenceserver.rb [hunk 4]
```diff
@@ -57,6 +58,83 @@ def init(config = {})
...
+          unless File.directory?(file_dir)
+            FileUtils.mkdir_p(file_dir)
+          end
+          FileUtils.mv([temp_stdout_file, temp_stderr_file][index], filename)
+        end
+      end
+
+      # If paths to write stdout and/or stderr to were not given, return the
+      # contents of stdout and/or stderr. Otherwise, return nil.
+      return_array = []
```
Comment (yeban) [inline lib/sequenceserver.rb]: I gave it some thought. Let's return if neither `:stdout_file` nor `:stderr_file` was specified.

File: lib/sequenceserver/database.rb [hunk 5]
```diff
@@ -113,16 +113,19 @@ def clear
...
+          out, err = sys(cmd)
+        rescue
+          fail BLAST_DATABASE_ERROR.new(cmd, err)
+        end
+        out.each_line do |line|
+          name = line.split('	')[0]
+          next if multipart_database_name?(name)
+          self << Database.new(*line.split('	'))
         end
+        errpat = /BLAST Database error/
```
Comment (yeban) [inline lib/sequenceserver/database.rb]: - Should check for errors and raise before parsing the output.
- Could you move running of the command and failure scenarios to a new method? Call it `blastdbmcmd`.
- Also, `throw_scan_error` is now redundant. You should remove it.

Comment (enjoyyourcopepods) [inline lib/sequenceserver/database.rb]: The rescue block with `fail BLAST_DATABASE_ERROR.new(cmd, err)` won't work, I think. When `sys` raises an exception nothing will be returned to `err`, therefore `err` cannot be passed to `BLAST_DATABASE_ERROR.new(cmd, err)`.

Comment (yeban) [inline lib/sequenceserver/database.rb]: Right. You will have to `rescue CommandFailed => e` and use `e.stderr`.

**Comment by yeban** [comment]
Thanks!

**Pull request merged**